chain/chain/src/chain.rs

```diff
@@ -1117,6 +1120,38 @@ impl Chain {
         Ok(header.signature().verify(header.hash().as_ref(), block_producer.public_key()))
     }
 
+    /// Optimization which checks if block with the given header can be reached from final head, and thus can be
+    /// finalized by this node.
+    /// If this is the case, returns Ok.
+    /// If we discovered that it is not the case, returns `Error::CannotBeFinalized`.
+    /// If too many parents were checked, returns Ok to avoid long delays.
+    fn check_if_finalizable(&self, header: &BlockHeader) -> Result<(), Error> {
+        let mut header = header.clone();
+        let final_head = self.final_head()?;
+        for _ in 0..NUM_PARENTS_TO_CHECK_FINALITY {
```

Comment:
I'm not sure about the need for this constant. This actually assumes that in the chain, we would never have a fork with 20 blocks. While this rarely happens, adding this assumption doesn't help much with the implementation in this function, so I don't think it should be added. the check that compares the block height with the final head height should be enough.

Reply:
> This actually assumes that in the chain, we would never have a fork with 20 blocks.

I don't see where we need this assumption. If there is a fork with >20 blocks, it means that we just can't apply our optimization and discard it here. Having constant 20 helps us to discard more useless blocks (see below).

> the check that compares the block height with the final head height should be enough.

I think the current check is helpful for scenarios as follows:

```
     /-> 2 -> 3 -> 5 -> 6 -> 7
   1
(final) 
     \-> 4 -> 8
```

* processing blocks (1) - (6) don't change final head, so no blocks can be discarded
* block (7) moves final head to (5)
* now block (8) can be discarded during preprocessing, but to know it, we need to look at its parent (4).

Reply:
I see. I still don't understand why we need to only check the last 20 blocks. What do you think about just changing the for loop to loop?

Reply:
I'm concerned about very long forks for which it can take a while to iterate over parents. Maybe there is some reason not to worry about it, but I don't see it.

Reply:
If I put the check to `preprocess_block`, then these tests are failing:
```
tests::challenges::challenges_new_head_prev
tests::simple_chain::blocks_at_height
```
They create some tree of blocks on chain, and the tests fail because some of blocks are being discarded.

Moved it to `Client::start_process_block`, because it looks like optimization on Client side, and I wouldn't make Chain tests assuming this optimization as well. Even after this, two other tests had to be changed.